### PR TITLE
Add support for both SystemAssigned and UserAssigned identities

### PR DESCRIPTION
**Pull request by primeroz**
towards https://github.com/giantswarm/giantswarm/issues/24578

### CODE REVIEW

**Comment by primeroz** [comment]
I think i hit a bug with MachinePools and SystemAssigned identities :( 

the `scope` for the machines in the machinepools is not rendered properly and so it gets set to 

`/providers/Microsoft.Authorization/roleAssignments/25f2ab3d-9ff7-41ff-908e-cd2de24dd9c5` 

instead of the correct one ( taken from a machinedeployment ) 

`/subscriptions/6b1f6e4a-6d0e-4aa4-9a5a-fbaca65a23b3/providers/Microsoft.Authorization/roleAssignments/d5017d3e-db8e-440a-9b22-77a0e5600468`

:man_facepalming:

**Comment by primeroz** [comment]
the changes being done in https://github.com/kubernetes-sigs/cluster-api-provider-azure/pull/2965 will also address this issue. 

I do actually have a super simple patch that would fix this specific problem and let me continue on the work towards the bigger issue 

```
diff --git a/azure/scope/machinepool.go b/azure/scope/machinepool.go
index 1d515a76..a917632c 100644
--- a/azure/scope/machinepool.go
+++ b/azure/scope/machinepool.go
@@ -590,11 +590,13 @@ func (m *MachinePoolScope) RoleAssignmentSpecs(principalID *string) []azure.Reso
    roles := make([]azure.ResourceSpecGetter, 1)
    if m.HasSystemAssignedIdentity() {
        roles[0] = &roleassignments.RoleAssignmentSpec{
-           Name:          m.AzureMachinePool.Spec.RoleAssignmentName,
-           MachineName:   m.Name(),
-           ResourceGroup: m.ResourceGroup(),
-           ResourceType:  azure.VirtualMachineScaleSet,
-           PrincipalID:   principalID,
+           Name:             m.AzureMachinePool.Spec.RoleAssignmentName,
+           MachineName:      m.Name(),
+           ResourceGroup:    m.ResourceGroup(),
+           ResourceType:     azure.VirtualMachineScaleSet,
+           Scope:            azure.GenerateSubscriptionScope(m.SubscriptionID()),
+           RoleDefinitionID: azure.GenerateContributorRoleDefinitionID(m.SubscriptionID()),
+           PrincipalID:      principalID,
        }
        return roles
    }
```

but i am not sure at the moment, with the status of our CAPZ fork, how to proceed to get a build with our custom patch out of latest upstream code

**Comment by primeroz** [comment]
we'll wait for the upstream PR

File: helm/cluster-azure/values.yaml [hunk 1]
```diff
@@ -45,12 +54,8 @@ machinePools:
...
     #   effect: "" # Valid effects are NoSchedule, PreferNoSchedule and NoExecute.
 
-# Only required until we have AZWI for CAPZ and only required on MC cluster
-attachCapzControllerIdentity: false
-
-# TODO: This is temporary so we can deploy clusters without dedicated UA Identity until the operator for WC clusters are ready
-# https://github.com/giantswarm/giantswarm/issues/25039
-enablePerClusterIdentity: false
+# Categorize this cluster as Management Cluster or Workload Cluster
+managementCluster: false
```
Comment (marians) [inline helm/cluster-azure/values.yaml:58]: Would be good to move this under /metadata

File: helm/cluster-azure/values.yaml [hunk 2]
```diff
@@ -31,6 +31,15 @@ defaults:
...
   evictionMinimumReclaim: imagefs.available=5%,memory.available=100Mi,nodefs.available=5%
+  identity:
+    # Only SystemAssigned or UserAssigned supported
+    type: SystemAssigned
+    # Not used yet https://github.com/kubernetes-sigs/cluster-api-provider-azure/pull/2965/files
+    scope: </subscriptions/${AZURE_SUBSCRIPTION_ID}/resourceGroups/${RESOURCE_GROUP_NAME}>
+    # Not used yet https://github.com/kubernetes-sigs/cluster-api-provider-azure/pull/2965/files
+    roleDefinitionID: </subscriptions/${AZURE_SUBSCRIPTION_ID}/providers/Microsoft.Authorization/roleDefinitions/b24988ac-6180-42a0-ab88-20f7382dd24c>
+    # Pass list of UA Identities to attach to nodes, if empty defaults -cp ,-nodes, -capz are used
+    userAssignedIdentities: []
```
Comment (marians) [inline helm/cluster-azure/values.yaml:42]: We should find a place in the new root level structure for these keys. I'm currently leaning towards /providerSpecific, as it seems to be specific to Azure.

See https://github.com/giantswarm/roadmap/issues/1990

Comment (marians) [reply helm/cluster-azure/values.yaml:42]: In case we want to hide this from customers, it should go to `/internal`.

**Pull request closed without merging**